src/dbnode/storage/flush.go

```diff
@@ -270,25 +296,103 @@ func (m *flushManager) indexFlush(
 			indexEnabled = indexOpts.Enabled()
 		)
 		if !indexEnabled {
 			continue
 		}
-		multiErr = multiErr.Add(ns.FlushIndex(indexFlush))
+		blockStarts, err := ns.FlushIndex(indexFlush)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+		}
+		nsResults[i] = blockStarts
 	}
 	multiErr = multiErr.Add(indexFlush.DoneIndex())
 
+	// Only write out empty index snapshots if the commit log rotation
+	// was successful we don't mind index flush errors since both
+	// index flush and writing empty index snapshots is best effort.
+	if rotateCommitlogSuccess {
+		multiErr = multiErr.Add(m.writeEmptyIndexSnapshots(
+			namespaces,
+			nsResults,
+			startTime,
+			rotatedCommitlogID,
+			snapshotID,
+		))
+	}
+
 	m.metrics.indexFlushDuration.Record(m.nowFn().Sub(start))
 	return multiErr.FinalError()
 }
 
+// NB(bodu): After a successful warm block flush, we must write an empty snapshot to disk.
+// If the DB node crashes before next snapshot is written to disk and we bootstrap the warm snapshot
+// into the now "sealed" block those warm bootstrapped segments will never get evicted from memory.
+// Since we are doing this after a successful flush, there is a small window where we can crash and
+// run into the above case. However, we'd rather deal w/ a one time extra mem than long bootstrap times
+// if we write an empty snapshot to disk and the node crashes before a successful warm flush.
+func (m *flushManager) writeEmptyIndexSnapshots(
+	namespaces []databaseNamespace,
+	nsResults [][]time.Time,
+	snapshotTime time.Time,
+	// We retain the commitlog ID of the prior data/index snapshot run during this flush
+	// since we haven't actually rotated the commit log again.
+	rotatedCommitlogID persist.CommitLogFile,
+	// We also need to retain the same snapshot ID because the data snapshot cleanup process
+	// relies on the latest snapshot metadata snapshot ID to perform clenaup.
+	snapshotID uuid.UUID,
+) error {
+	snapshotPersist, err := m.pm.StartSnapshotPersist(snapshotID)
+	if err != nil {
+		return err
+	}
+
+	// Since index warm flush is best effort, we will attempt to be best effort here as well.
+	multiErr := xerrors.NewMultiError()
+	for i, ns := range namespaces {
+		results := nsResults[i]
+		if len(results) == 0 {
+			continue
+		}
+
+		allShards := make(map[uint32]struct{})
+		for _, dbShard := range ns.OwnedShards() {
+			allShards[dbShard.ID()] = struct{}{}
+		}
+		for _, blockStart := range results {
+			prepareOpts := persist.IndexPrepareSnapshotOptions{
+				IndexPrepareOptions: persist.IndexPrepareOptions{
+					NamespaceMetadata: ns.Metadata(),
+					BlockStart:        blockStart,
+					FileSetType:       persist.FileSetSnapshotType,
+					Shards:            allShards,
+					IndexVolumeType:   idxpersist.SnapshotWarmIndexVolumeType,
+				},
+				SnapshotTime: snapshotTime,
+			}
+			prepared, snapshotErr := snapshotPersist.PrepareIndexSnapshot(prepareOpts)
+			if snapshotErr == nil {
+				_, snapshotErr = prepared.Close()
```

Comment:
Do we want to `continue` here if `snapshotErr` is nil?

Reply:
Or is multiErr.Add smart enough to ignore nils?

Reply:
`multiErr.Add()` ignores `nil`s 
```
// Add adds an error returns a new MultiError object.
func (e MultiError) Add(err error) MultiError {
        if err == nil {
                return e
        }
```